### PR TITLE
CORDA-3754 Make TransactionBuilder's input references deep copied on copy.

### PR DESCRIPTION
**Pull request by xavierlepretre**
On `TransactionBuilder.copy`, I made sure that `referenceStates` are deep copied like the other lists. [Before](https://github.com/corda/corda/blob/963de40902c836d17d2aabe714bbdfacc62c096e/core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt#L103) / [After](https://github.com/xavierlepretre/corda/blob/9396ee95512a51e7def962121c58d40432c157d3/core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt#L103).

I discovered it when using the [`TransactionDSLInterpreter.tweak` function](https://github.com/corda/corda/blob/963de40902c836d17d2aabe714bbdfacc62c096e/testing/test-utils/src/main/kotlin/net/corda/testing/dsl/TransactionDSLInterpreter.kt#L207). I discovered that the reference state I added inside `tweak { }` was still in the `TransactionDSL` outside the `tweak` scope.

# PR Checklist:

- [x] Have you run the unit, integration and smoke tests as described [here](https://docs.corda.net/head/testing.html)?
- [ ] If you added public APIs, did you write the JavaDocs?
- [ ] If the changes are of interest to application developers, have you added them to the [changelog](https://docs.corda.net/head/changelog.html) (`/docs/source/changelog.rst`), and potentially the [release notes](https://docs.corda.net/head/release-notes.html) (`/docs/source/release-notes.rst`)?
- [x] If you are contributing for the first time, please read the [contributor agreement](https://docs.corda.net/head/contributing.html) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://docs.corda.net/head/contributing.html#merging-the-changes-back-into-corda).

Thanks for your code, it's appreciated! :)

### CODE REVIEW

**Comment by xavierlepretre** [comment]
My PR is in accordance with the Developer's Certificate of Origin.

**Comment by davidrapacchiale** [comment]
Thanks @xavierlepretre for submitting this PR. I have included my colleagues @rick-r3 and @mnesbit as reviewers and they'll get back to you (time permitting)

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 1]
```diff
@@ -813,19 +813,19 @@ open class TransactionBuilder(
         this.privacySalt = privacySalt
     }
 
-    /** Returns an immutable list of input [StateRef]s. */
+    /** Returns a mutable copy of the list of input [StateRef]s. */
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: Although the implementation class is mutable the interface is intentionally immutable. Can we revert these comment changes.

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 2]
```diff
@@ -100,7 +100,7 @@ open class TransactionBuilder(
                 commands = ArrayList(commands),
                 window = window,
                 privacySalt = privacySalt,
-                references = references,
+                references = ArrayList(references),
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt:103]: this does make sense

**Review by mnesbit — COMMENTED** [review]
Not a great fan of the transaction DSL, but I agree the reference states should be deep copied. However, the comments changes are not correct as the public interface is intentionally immutable.

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt (hunk 1 above)
Comment (xavierlepretre) [reply core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: I can sure do it. 
I reckon we should also ensure that the implementation returns an immutable list, and not a mutable one as is proven in the test:

```kotlin
(builder.referenceStates() as ArrayList).add(referenceStateRef2)
```
The idea is to protect the developer who took the interface at face value and failed to look at the implementation.

Comment (mnesbit) [reply core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: I think we considered that in the past, but can get a bit challenging for serialization. Certainly a possibility for the future.

**Review by mnesbit — COMMENTED** [review]
The timeline for OS4.5 starting our QA is soon, so be good to get the comments reverted and we could get this in to 4.5 cut, otherwise will have to be pushed onto 4.6.

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt (hunk 1 above)
Comment (xavierlepretre) [reply core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: Done.

**Review by mnesbit — APPROVED** [review]
OK

**Comment by nargas-ritu** [comment]
Minor change and merging as it is to be included in RC01 for 4.5. pr_merge was breaking due to flaky tests.

**Pull request merged**